lib/charms/rolling_ops/v0/rollingops.py

```diff
@@ -0,0 +1,353 @@
+"""A library to enable charms to implement "rolling" or "serialized" operations. E.g., a
+rolling restart.
+
+You may use this library directly, or extend it to customize behavior.
+
+For example, in order to implement a rolling restart, a charm author would need to add the
+following lines of code, to the following files (note the consistent use of the name "restart")
+to namespace this particular rolling op:
+
+Add a peer relation to `metadata.yaml`:
+```yaml
+peers:
+    restart:
+        interface: rolling_op
+```
+
+To enable a human operator to trigger the restat, add an action to `actions.yaml`:
+```yaml
+restart:
+  description: Restarts a service
+  params: {}
+```
+
+Import the library, and enable it by doing the following:
+
+1. Override the Charm's `on` property with the RollingEvents class from the library.
+2. Add a method containing code to be executed to the Charm object. Here, we call this
+method `_restart`.
+3. Initialize a RollingOpsManager class, passing in the Charm, name of the peer relation,
+restart handler, and action.
+
+src/charm.py
+```python
+# ...
+from charms.rolling_ops.v0.rollingops import RollingOpsManager, RollingEvents
+# ...
+
+class SomeCharm(...):
+    # ...
+    on = RollingEvents
```

Comment:
I feel like we can take care of this for the developer in the library init? We're passing `self` into the library, so the library could do a `self.charm.on = RollingEvents()`. My only concern is if that might override any *other* custom event sources, so perhaps `self.charm.on['rolling_ops'] = RollingEvents()` to keep them 'namespaced'?

Also I think it should be `RollingEvents()` not `RollingEvents`, right?

Reply:
The risk of silently clobbering is why I left this here (as per our docs), rather than trying to fold it into the library.

And it is RollingEvents, with no call. The class is basically just a list of allowed events. Initializing it doesn't do anything.

As we use more libraries, we need a better way of doing this, but I'm not ready to tackle that quite yet. For now, this at least makes it clear that you will be creating problems if you try to attach the events from two different libraries to on.

Reply:
Edit: nevermind. The call to init the class is necessary. Per my own code, written below :-)

But it is Events, plural, not Event, singular.